runtime/near-vm-runner/src/wasmer1_runner.rs

```diff
@@ -94,19 +96,58 @@ impl IntoVMError for wasmer::RuntimeError {
-            _ => {
-                if let Some(trap_code) = trap_code {
-                    // A trap
-                    VMError::FunctionCallError(FunctionCallError::Wasmer1Trap(
-                        trap_code.to_string(),
-                    ))
-                } else {
-                    // A general error
-                    VMError::FunctionCallError(FunctionCallError::WasmerRuntimeError(error_msg))
-                }
-            }
+            Ok(e) => return (&e).into(),
+            _ => {}
         }
+        let error = match trap_code {
+            Some(TrapCode::StackOverflow) => FunctionCallError::WasmTrap(WasmTrap::StackOverflow),
+            Some(TrapCode::HeapSetterOutOfBounds) => {
+                FunctionCallError::WasmTrap(WasmTrap::MemoryOutOfBounds)
+            }
+            Some(TrapCode::HeapAccessOutOfBounds) => {
+                FunctionCallError::WasmTrap(WasmTrap::MemoryOutOfBounds)
+            }
+            Some(TrapCode::HeapMisaligned) => {
+                FunctionCallError::WasmTrap(WasmTrap::MisalignedAtomicAccess)
+            }
+            Some(TrapCode::TableSetterOutOfBounds) => {
+                FunctionCallError::WasmTrap(WasmTrap::MemoryOutOfBounds)
+            }
+            Some(TrapCode::TableAccessOutOfBounds) => {
+                FunctionCallError::WasmTrap(WasmTrap::MemoryOutOfBounds)
+            }
+            Some(TrapCode::OutOfBounds) => FunctionCallError::WasmTrap(WasmTrap::MemoryOutOfBounds),
+            Some(TrapCode::IndirectCallToNull) => {
+                FunctionCallError::WasmTrap(WasmTrap::IncorrectCallIndirectSignature)
+            }
+            Some(TrapCode::BadSignature) => {
+                FunctionCallError::WasmTrap(WasmTrap::IncorrectCallIndirectSignature)
+            }
+            Some(TrapCode::IntegerOverflow) => {
+                FunctionCallError::WasmTrap(WasmTrap::IllegalArithmetic)
+            }
+            Some(TrapCode::IntegerDivisionByZero) => {
+                FunctionCallError::WasmTrap(WasmTrap::IllegalArithmetic)
+            }
+            Some(TrapCode::BadConversionToInteger) => {
+                FunctionCallError::WasmTrap(WasmTrap::IllegalArithmetic)
+            }
+            Some(TrapCode::UnreachableCodeReached) => {
+                // Not FunctionCallError::WasmTrap(WasmTrap::Unreachable) for Wasmer 0.x compatibility.
+                FunctionCallError::WasmUnknownError
+            }
+            Some(TrapCode::Interrupt) => {
+                FunctionCallError::Nondeterministic("Wasmer interrupt".to_string())
+            }
+            Some(TrapCode::UnalignedAtomic) => {
+                FunctionCallError::WasmTrap(WasmTrap::MisalignedAtomicAccess)
+            }
+            Some(TrapCode::VMOutOfMemory) => {
+                FunctionCallError::Nondeterministic("Wasmer out of memory".to_string())
+            }
+            None => FunctionCallError::Nondeterministic(error_msg),
```

Comment:
This is incorrect, this only happens when wasmer returns a Generic Error or User Error that failed to downcast to VMLogicError: https://github.com/wasmerio/wasmer/blob/9e0122e1f7bdb6fedf5fc17cc9a3a0fda30031a0/lib/engine/src/trap/error.rs#L18
And Generic Error & User Error must come from we construct it manually in contract, which is unreachable, because we only construct VMLogicError from there. So this is not an Nondeterministic, should be debug_assert!(false) or unreachable!()

Reply:
Reworked code around this place.